[caffe2] Fix pybind11 native python link error

loren-arthur opened the pull request:
Summary:
Currently, we define some C++ functions in one C++ Python extension
which are used by another.  This happens to work, but isn't guaranteed to.
This diff moves these functions to a separate C++ library rule to fix this.

Test Plan: CI

Differential Revision: D42552515

Skylion007 commented:
I see no reason why the original code shouldn't be guaranteed to work given that they proper linking to each other and that the ABI / pybind11 version is the same? This code change is probably better for other reasons though.

loren-arthur committed:
[caffe2] Fix pybind11 native python link error (#92325)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/92325

Currently, we define some C++ functions in one C++ Python extension
which are used by another.  This happens to work, but isn't guaranteed to.
This diff moves these functions to a separate C++ library rule to fix this.

Test Plan: CI

Reviewed By: andrewjcg

Differential Revision: D42552515

fbshipit-source-id: 774796e3af65bf9892205566d02a4c1655cd13ec
